state/tree.go

```diff
@@ -243,42 +243,54 @@ func (t *tree) debugPointer(ps []*hamt.Pointer) {
 	}
 }
 
-// GetAllActors returns a slice of all actors in the StateTree, t.
-func GetAllActors(t Tree) ([]string, []*actor.Actor) {
-	st := t.(*tree)
-
-	return st.getActorsFromPointers(st.root.Pointers)
+// GetAllActorsResult is the struct returned via a channel by the GetAllActors
+// method. This struct contains only an address string and the actor itself.
+type GetAllActorsResult struct {
+	Address string
+	Actor   *actor.Actor
 }
 
-// GetAllActorsFromStoreFunc is a function with the signature of GetAllActorsFromStore
-type GetAllActorsFromStoreFunc = func(context.Context, *hamt.CborIpldStore, cid.Cid) ([]string, []*actor.Actor, error)
-
-// GetAllActorsFunc is a function with the signature of GetAllActors
-type GetAllActorsFunc = func(t Tree) ([]string, []*actor.Actor)
+// GetAllActors returns a channel which provides all actors in the StateTree, t.
+func GetAllActors(ctx context.Context, t Tree) <-chan GetAllActorsResult {
+	st := t.(*tree)
+	out := make(chan GetAllActorsResult)
+	go func() {
+		st.getActorsFromPointers(ctx, out, st.root.Pointers)
+		close(out)
+	}()
+	return out
+}
 
-// GetAllActorsFromStore loads a StateTree and returns arrays of addresses and their corresponding actors.
-// Third returned value is any error that occurred when loading.
-func GetAllActorsFromStore(ctx context.Context, store *hamt.CborIpldStore, stateRoot cid.Cid) ([]string, []*actor.Actor, error) {
+// GetAllActorsFromStore loads a StateTree and returns a channel with addresses
+// and their corresponding actors. The second returned value is any error that
+// occurred when loading.
+func GetAllActorsFromStore(ctx context.Context, store *hamt.CborIpldStore, stateRoot cid.Cid) (<-chan GetAllActorsResult, error) {
 	st, err := LoadStateTree(ctx, store, stateRoot, nil)
 	if err != nil {
-		return nil, nil, err
+		return nil, err
 	}
-
-	addrs, actors := GetAllActors(st)
-	return addrs, actors, nil
+	return GetAllActors(ctx, st), nil
 }
 
 // NOTE: This extracts actors from pointers recursively. Maybe we shouldn't recurse here.
-func (t *tree) getActorsFromPointers(ps []*hamt.Pointer) (addresses []string, actors []*actor.Actor) {
+func (t *tree) getActorsFromPointers(ctx context.Context, out chan<- GetAllActorsResult, ps []*hamt.Pointer) {
 	for _, p := range ps {
 		for _, kv := range p.KVs {
 			var a actor.Actor
 			if err := hackTransferObject(kv.Value, &a); err != nil {
 				panic(err) // uhm, ignoring errors is bad
 			}
 
-			addresses = append(addresses, kv.Key)
-			actors = append(actors, &a)
+			result := GetAllActorsResult{
+				Address: kv.Key,
+				Actor:   &a,
+			}
+
+			select {
+			case <-ctx.Done():
```

Comment:
maybe add a note that select is non deterministic so there is some chance that if receiver is waiting we will send them a result even when the context has been canceled. to fix this we could put an if ctx.err != nil return above this select, and keep the select as is, but dunno if that is worth it or just confusing... cc @anorth